accord-core/src/main/java/accord/primitives/Keys.java

```diff
@@ -0,0 +1,449 @@
+     * Count the number of keys matching the predicate and intersecting with the given ranges.
+     * If terminateAfter is greater than 0, the method will return once terminateAfter matches are encountered
+     */
+    @Inline
+    public <V> V foldl(KeyRanges rs, IndexedFold<Key, V> fold, V accumulator)
+    {
+        int ai = 0, ri = 0;
+        while (true)
+        {
+            long ari = rs.findNextIntersection(ri, this, ai);
+            if (ari < 0)
+                break;
+
+            ai = (int)(ari >>> 32);
+            ri = (int)ari;
+            KeyRange range = rs.get(ri);
+            int nextai = range.higherKeyIndex(this, ai + 1, keys.length);
+            while (ai < nextai)
+            {
+                accumulator = fold.apply(ai, get(ai), accumulator);
+                ++ai;
+            }
+        }
+
+        return accumulator;
+    }
+
+    @Inline
+    public long foldl(KeyRanges rs, IndexedFoldToLong<Key> fold, long param, long initialValue, long terminalValue)
+    {
+        int ai = 0, ri = 0;
+        done: while (true)
+        {
+            long ari = rs.findNextIntersection(ri, this, ai);
+            if (ari < 0)
+                break;
+
+            ai = (int)(ari >>> 32);
+            ri = (int)ari;
+            KeyRange range = rs.get(ri);
+            int nextai = range.higherKeyIndex(this, ai + 1, keys.length);
+            while (ai < nextai)
+            {
+                initialValue = fold.apply(ai, get(ai), param, initialValue);
+                if (initialValue == terminalValue)
+                    break done;
+                ++ai;
+            }
+        }
+
+        return initialValue;
+    }
+
+    /**
+     * A fold variation permitting more efficient operation over indices only, by providing ranges of matching indices
+     */
+    @Inline
+    public long rangeFoldl(KeyRanges rs, IndexedRangeFoldToLong fold, long param, long initialValue, long terminalValue)
+    {
+        int ai = 0, ri = 0;
+        while (true)
+        {
+            long ari = rs.findNextIntersection(ri, this, ai);
+            if (ari < 0)
+                break;
+
+            ai = (int)(ari >>> 32);
+            ri = (int)ari;
+            KeyRange range = rs.get(ri);
+            int nextai = range.higherKeyIndex(this, ai + 1, keys.length);
+            initialValue = fold.apply(ai, nextai, param, initialValue);
+            if (initialValue == terminalValue)
+                break;
+            ai = nextai;
+        }
+
+        return initialValue;
+    }
+
+    @Inline
+    public long foldl(IndexedFoldToLong<Key> fold, long param, long initialValue, long terminalValue)
+    {
+        for (int i = 0; i < keys.length; i++)
+        {
+            initialValue = fold.apply(i, keys[i], param, initialValue);
+            if (terminalValue == initialValue)
+                return initialValue;
+        }
+        return initialValue;
+    }
+
+    public boolean intersects(KeyRanges ranges, Keys keys)
```

Comment:
not used, if we get rid of then can also drop `foldlIntersect(KeyRanges rs, Keys intersect`